up_esb/execution/executor.py

```diff
@@ -0,0 +1,322 @@
+"""Executor for executing tasks."""
+from threading import Condition, Lock, Thread
+from typing import NamedTuple
+
+import networkx as nx
+from unified_planning.plans import Plan, SequentialPlan, TimeTriggeredPlan
+from unified_planning.shortcuts import EndTiming, StartTiming, TimeInterval
+
+from up_esb.status import ActionNodeStatus, ConditionStatus
+
+# Action Result
+ActionResult = NamedTuple(
+    "ActionResult",
+    [
+        ("precondition_status", ConditionStatus),
+        ("action_status", ActionNodeStatus),
+        ("postcondition_status", ConditionStatus),
+        ("result", object),
+    ],
+)
+
+
+class TaskTracker:
+    """Track the amount of tasks that is being executed."""
+
+    def __init__(self):
+        self._lock = Lock()
+        self._condition = Condition(self._lock)
+        self._tasks = 0
+
+    def __enter__(self):
+        with self._lock:
+            self._tasks += 1
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        with self._lock:
+            self._tasks -= 1
+            self._condition.notify()
+
+    def wait(self):
+        """Wait until all tasks are finished."""
+        with self._lock:
+            while self._tasks > 0:
+                self._condition.wait()
+
+
+class TaskExecutor:
+    """Base class for task executors."""
+
+    def __init__(self):
+        self._lock = Lock()
+        self._condition = Condition(self._lock)
+        self._task = None
+        self._thread = None
+        self._task_tracker = None
+
+
+class ActionExecutor:
+    """Base class for action executors."""
+
+    def __init__(self, dependency_graph: nx.DiGraph, options: dict):
+        self._dependency_graph = dependency_graph
+        # Options
+        self._options = options
+        self._dry_run = options.get("dry_run", False)
+        self._verbose = options.get("verbose", False)
+        self._context = None
+        self._action = None
+        self._node_name = None
+        self._parameters = None
+
+        self._supported_plan_kind = None
+
+    def get_executor(self, plan: Plan):
```

Comment:
Here, I find it confusing in the examples that  `get_executor` is a method. So in the examples and tests you first create an `ActionExecutor` which is then updated to the subclass by calling this method.